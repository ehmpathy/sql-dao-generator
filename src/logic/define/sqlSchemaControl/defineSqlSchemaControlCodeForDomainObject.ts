import { UnexpectedCodePathError } from '@ehmpathy/error-fns';
import { snakeCase } from 'change-case';

import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { UnexpectedCodePathDetectedError } from '../../UnexpectedCodePathDetectedError';

const getMostCommonPrefixRedactedTableNameToken = (input: {
  sqlSchemaRelationship: SqlSchemaToDomainObjectRelationship;
  propertyRelationship: SqlSchemaToDomainObjectRelationship['properties'][number];
}) => {
  if (!input.propertyRelationship.sqlSchema.reference)
    throw new UnexpectedCodePathError(
      'can not getMostCommonPrefixRedactedTableNameToken of non reference property',
      { input },
    );

  // define the names of the tables
  const tableNameSource = input.sqlSchemaRelationship.name.sqlSchema;
  const tableNameReferenced = snakeCase(
    input.propertyRelationship.sqlSchema.reference.of.name,
  );

  // determine the most common prefix between the two, if any
  const wordsInEntityReferenceTableName = tableNameSource.split('_');
  const mostCommonPrefixWords = (() => {
    const commonPrefixWords: string[] = [];
    for (const commonPrefixCandidate of wordsInEntityReferenceTableName) {
      const doesCandidateContributeToCommonPrefix =
        tableNameReferenced.startsWith(
          [...commonPrefixWords, commonPrefixCandidate].join('_'),
        );
      if (!doesCandidateContributeToCommonPrefix) break;
      commonPrefixWords.push(commonPrefixCandidate);
    }
    return commonPrefixWords;
  })();
  const mostCommonPrefix = mostCommonPrefixWords.join('_');

  // establish the join table name with the prefix removed
  const referencedTableNameWithPrefixDeduped = tableNameReferenced.replace(
    new RegExp(`^${mostCommonPrefix}_`), // todo: sync this logic to the sql-schema-generator
    '',
  );

  // return it
  return referencedTableNameWithPrefixDeduped;
};

export const defineSqlSchemaControlCodeForDomainObject = ({
  sqlSchemaRelationship,
}: {
  sqlSchemaRelationship: SqlSchemaToDomainObjectRelationship;
}) => {
  // derive some metadata
  const hasArrayProperties = sqlSchemaRelationship.properties.some(
    (propertyRelationship) => propertyRelationship.sqlSchema.isArray,
  );
  const hasUpdatableProperties = sqlSchemaRelationship.properties.some(
    (propertyRelationship) => propertyRelationship.sqlSchema.isUpdatable,
  );

  // track all of the relpaths for resources
  const resourceRelpaths: string[] = [];

  // define the base table
  resourceRelpaths.push(`./tables/${sqlSchemaRelationship.name.sqlSchema}.sql`);

  // define any mapping tables referencing the base table
  sqlSchemaRelationship.properties.forEach((propertyRelationship) => {
    if (!propertyRelationship.sqlSchema.isArray) return;
    if (propertyRelationship.sqlSchema.isUpdatable) return; // only ones that join to static table
    if (!propertyRelationship.sqlSchema.reference) {
      const endsWithUuidSuffix = new RegExp(/_uuids$/).test(
        propertyRelationship.sqlSchema.name,
      );
      if (!endsWithUuidSuffix)
        throw new UnexpectedCodePathDetectedError({
          reason:
            'expected sql-schema-generator to only allow _uuid suffixed property names to be arrays',
          domainObjectPropertyName: propertyRelationship.domainObject?.name,
          domainObjectName: sqlSchemaRelationship.name.domainObject,
        });
      return resourceRelpaths.push(
        `./tables/${
          sqlSchemaRelationship.name.sqlSchema
        }_to_${propertyRelationship.sqlSchema.name.replace(
          /_uuids$/,
          '_uuid',
        )}.sql`,
      );
    }
    return resourceRelpaths.push(
      `./tables/${
        sqlSchemaRelationship.name.sqlSchema
      }_to_${getMostCommonPrefixRedactedTableNameToken({
        sqlSchemaRelationship,
        propertyRelationship,
      })}.sql`,
    );
  });

  // define the version table, if there are any updatable properties
  if (hasUpdatableProperties)
    resourceRelpaths.push(
      `./tables/${sqlSchemaRelationship.name.sqlSchema}_version.sql`,
    );

  // define any mapping tables referencing the version table, if there are any updatable properties
  if (hasUpdatableProperties)
    sqlSchemaRelationship.properties.forEach((propertyRelationship) => {
      if (!propertyRelationship.sqlSchema.isArray) return;
      if (!propertyRelationship.sqlSchema.isUpdatable) return; // only ones that join to static table
      if (!propertyRelationship.sqlSchema.reference) {
        const endsWithUuidSuffix = new RegExp(/_uuids$/).test(
          propertyRelationship.sqlSchema.name,
        );
        if (!endsWithUuidSuffix)
          throw new UnexpectedCodePathDetectedError({
            reason:
              'expected sql-schema-generator to only allow _uuid suffixed property names to be arrays',
            domainObjectPropertyName: propertyRelationship.domainObject?.name,
            domainObjectName: sqlSchemaRelationship.name.domainObject,
          });
        return resourceRelpaths.push(
          `./tables/${
            sqlSchemaRelationship.name.sqlSchema
          }_version_to_${propertyRelationship.sqlSchema.name.replace(
            /_uuids$/,
            '_uuid',
          )}.sql`,
        );
      }
      return resourceRelpaths.push(
        `./tables/${
          sqlSchemaRelationship.name.sqlSchema
        }_version_to_${getMostCommonPrefixRedactedTableNameToken({
          sqlSchemaRelationship,
          propertyRelationship,
        })}.sql`,
      );
    });

  // define the current version pointer table, if there are any updatable properties
  if (hasUpdatableProperties)
    resourceRelpaths.push(
      `./tables/${sqlSchemaRelationship.name.sqlSchema}_cvp.sql`,
    );

  // define the "current" view, if there are any updatable properties or array properties
  if (hasUpdatableProperties || hasArrayProperties)
    resourceRelpaths.push(
      `./views/view_${sqlSchemaRelationship.name.sqlSchema}_current.sql`,
    );

  // define the "hydrated" view, always
  resourceRelpaths.push(
    `./views/view_${sqlSchemaRelationship.name.sqlSchema}_hydrated.sql`,
  );

  // define the upsert function
  resourceRelpaths.push(
    `./functions/upsert_${sqlSchemaRelationship.name.sqlSchema}.sql`,
  );

  // define the full code for this domain entity, now that we have all of the paths to the resources defined, in order
  const code = [
    `# ${sqlSchemaRelationship.name.sqlSchema}`,
    ...resourceRelpaths.map((relpath) =>
      ['- type: resource', `  path: ${relpath}`].join('\n'),
    ),
  ].join('\n');

  // return the code
  return code;
};
