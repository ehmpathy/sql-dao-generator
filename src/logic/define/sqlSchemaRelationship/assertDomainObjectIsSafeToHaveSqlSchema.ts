import { DomainObjectMetadata, DomainObjectVariant } from 'domain-objects-metadata';
import { UserInputError } from '../../UserInputError';

export const assertDomainObjectIsSafeToHaveSqlSchema = ({ domainObject }: { domainObject: DomainObjectMetadata }) => {
  // make sure that entities have "unique" and "updatable" defined
  if (domainObject.extends === DomainObjectVariant.DOMAIN_ENTITY) {
    if (!domainObject.decorations.unique?.length)
      throw new UserInputError({
        reason:
          "domain entities must have at least one 'unique' property defined in order for a schema to be generated.",
        domainObjectName: domainObject.name,
        potentialSolution:
          "(note: if its not unique on any natural keys, please specify the 'uuid' as it's only unique property. e.g., `['uuid']`) ",
      });
    if (!Array.isArray(domainObject.decorations.updatable))
      throw new UserInputError({
        reason:
          "domain entities must have their 'updatable' properties defined in order for a schema to be generated. ",
        domainObjectName: domainObject.name,
        potentialSolution:
          '(note: if it has no updatable properties, please specify this with an empty array. e.g., `[]`) ',
      });
  }

  // make sure that value objects do not have unique or updatable defined
  if (domainObject.extends === DomainObjectVariant.DOMAIN_VALUE_OBJECT) {
    if (domainObject.decorations.unique)
      throw new UserInputError({
        reason:
          "domain value objects must _not_ have their 'unique' properties specified. value objects are unique on all of their properties by definition.",
        domainObjectName: domainObject.name,
      });
    if (domainObject.decorations.updatable)
      throw new UserInputError({
        reason:
          "domain value objects must _not_ have any 'updatable' properties specified. value objects do not have any updatable properties by definition.",
        domainObjectName: domainObject.name,
      });
  }

  // make sure that events have "unique" defined
  if (domainObject.extends === DomainObjectVariant.DOMAIN_EVENT) {
    if (!domainObject.decorations.unique?.length)
      throw new UserInputError({
        reason: "domain events must have at least one 'unique' property defined in order for a schema to be generated",
        domainObjectName: domainObject.name,
      });
  }
};
