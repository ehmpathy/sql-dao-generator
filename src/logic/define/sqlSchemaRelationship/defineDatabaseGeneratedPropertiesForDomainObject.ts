import { DomainObjectMetadata } from 'domain-objects-metadata';
import { isPresent } from 'type-fns';

import { SqlSchemaPropertyMetadata } from '../../../domain/objects/SqlSchemaPropertyMetadata';

export const defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject = ({
  domainObject,
}: {
  domainObject: DomainObjectMetadata;
}) => {
  const isUniqueOnUuid = domainObject.decorations.unique?.includes('uuid');
  const hasUpdatableProperties =
    (domainObject.decorations.updatable?.length ?? 0) > 0;

  // define which sql schema properties are generated by the db
  return [
    new SqlSchemaPropertyMetadata({
      name: 'id',
      isDatabaseGenerated: true,
      isArray: false,
      isNullable: false,
      isUpdatable: false,
      reference: null,
    }),
    !isUniqueOnUuid
      ? new SqlSchemaPropertyMetadata({
          name: 'uuid',
          isDatabaseGenerated: true,
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          reference: null,
        })
      : null, // "uuid" is only autogenerated if the domain object is not unique on it
    new SqlSchemaPropertyMetadata({
      name: 'created_at',
      isDatabaseGenerated: true,
      isArray: false,
      isNullable: false,
      isUpdatable: false,
      reference: null,
    }),
    hasUpdatableProperties
      ? new SqlSchemaPropertyMetadata({
          name: 'effective_at',
          isDatabaseGenerated: true,
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          reference: null,
        })
      : null, // domain objects only have "effective_at" defined if they're updatable
    hasUpdatableProperties
      ? new SqlSchemaPropertyMetadata({
          name: 'updated_at',
          isDatabaseGenerated: true,
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          reference: null,
        })
      : null, // domain objects only have "updated_at" defined if they're updatable
  ].filter(isPresent);
};
