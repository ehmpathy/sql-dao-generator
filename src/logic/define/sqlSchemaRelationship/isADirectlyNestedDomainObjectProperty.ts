import {
  DomainObjectPropertyMetadata,
  isDomainObjectArrayProperty,
  isDomainObjectReferenceProperty,
} from 'domain-objects-metadata';

import { SqlSchemaPropertyMetadata } from '../../../domain';
import { isAUserDefinedDomainObjectProperty } from './isAUserDefinedDomainObjectProperty';

/**
 * specifies whether this domain object property is a reference to another domain object directly by nesting (in array or solo)
 */
export const isADirectlyNestedDomainObjectProperty = (propertyRelationship: {
  domainObject: DomainObjectPropertyMetadata | null;
  sqlSchema: SqlSchemaPropertyMetadata;
}): propertyRelationship is {
  domainObject: DomainObjectPropertyMetadata;
  sqlSchema: SqlSchemaPropertyMetadata;
} => {
  if (!isAUserDefinedDomainObjectProperty(propertyRelationship)) return false; // if not user defined, then its def not a reference
  if (isDomainObjectReferenceProperty(propertyRelationship.domainObject))
    return true; // if its directly nested, its directly nested
  if (
    isDomainObjectArrayProperty(propertyRelationship.domainObject) &&
    isDomainObjectReferenceProperty(propertyRelationship.domainObject.of)
  )
    return true; // if its an array of directly nested objects, still true
  return false; // otherwise, false
};
