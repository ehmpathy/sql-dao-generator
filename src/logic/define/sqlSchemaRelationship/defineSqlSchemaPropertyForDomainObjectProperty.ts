import { camelCase, snakeCase } from 'change-case';
import { DomainObjectMetadata, DomainObjectPropertyMetadata, DomainObjectPropertyType } from 'domain-objects-metadata';

import { SqlSchemaPropertyMetadata } from '../../../domain/objects/SqlSchemaPropertyMetadata';
import { SqlSchemaReferenceMethod } from '../../../domain/objects/SqlSchemaReferenceMetadata';
import { UnexpectedCodePathDetectedError } from '../../UnexpectedCodePathDetectedError';
import { UserInputError } from '../../UserInputError';
import { defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject } from './defineDatabaseGeneratedPropertiesForDomainObject';
import { defineSqlSchemaReferenceForDomainObjectProperty } from './defineSqlSchemaReferenceForDomainObjectProperty';

export const defineSqlSchemaPropertyForDomainObjectProperty = ({
  property,
  domainObject,
  allDomainObjects,
}: {
  property: DomainObjectPropertyMetadata;
  domainObject: DomainObjectMetadata;
  allDomainObjects: DomainObjectMetadata[];
}): SqlSchemaPropertyMetadata => {
  // sanity check that property name is in camel case
  if (camelCase(property.name) !== property.name)
    throw new UserInputError({
      reason: 'property names must be camel case',
      domainObjectName: domainObject.name,
      domainObjectPropertyName: property.name,
    });

  // sanity check that this is not called for a database generated property (since that should be handled elsewhere)
  const databaseGeneratedProperties = defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject({ domainObject });
  if (databaseGeneratedProperties.map(({ name }) => name).includes(snakeCase(property.name)))
    throw new UnexpectedCodePathDetectedError({
      reason: '`defineSqlSchemaPropertyForDomainObjectProperty was called for an db-generated property',
      domainObjectName: domainObject.name,
      domainObjectPropertyName: property.name,
    });

  // define data about this property
  const isUpdatable = domainObject.decorations.updatable?.includes(property.name) ?? false;
  const isNullable = !!property.nullable;
  const isArray = property.type === DomainObjectPropertyType.ARRAY;

  // determine if this is a reference of another domain object
  const reference = defineSqlSchemaReferenceForDomainObjectProperty({
    property,
    domainObject,
    allDomainObjects,
  });

  // define the schema property name
  const schemaPropertyName = (() => {
    const baseSchemaPropertyName = snakeCase(property.name);
    if (reference) {
      if (isArray) {
        if (reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING)
          return `${baseSchemaPropertyName.replace(/s$/, '')}_ids`; // suffix w/ ids, since its an array of fks
        if (reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID)
          return `${baseSchemaPropertyName.replace(/_uuids$/, '')}_ids`; // suffix w/ ids, since its an array of fks
      }
      if (reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING) return `${baseSchemaPropertyName}_id`; // suffix w/ id, since its a fk
      if (reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID)
        return `${baseSchemaPropertyName.replace(/_uuid$/, '')}_id`; // suffix w/ id, since its a fk
    }
    return baseSchemaPropertyName; // otherwise, its just the regular name, no modifications
  })();

  return new SqlSchemaPropertyMetadata({
    name: schemaPropertyName,
    isUpdatable,
    isNullable,
    isArray,
    isDatabaseGenerated: false, // this method is only called for properties that are not database generated (as checked above)
    reference,
  });
};
