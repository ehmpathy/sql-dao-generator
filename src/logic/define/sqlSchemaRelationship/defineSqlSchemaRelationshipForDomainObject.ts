import { snakeCase } from 'change-case';
import {
  DomainObjectMetadata,
  DomainObjectPropertyType,
  DomainObjectVariant,
} from 'domain-objects-metadata';

import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { UnexpectedCodePathDetectedError } from '../../UnexpectedCodePathDetectedError';
import { UserInputError } from '../../UserInputError';
import { assertDomainObjectIsSafeToHaveSqlSchema } from './assertDomainObjectIsSafeToHaveSqlSchema';
import { defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject } from './defineDatabaseGeneratedPropertiesForDomainObject';
import { defineSqlSchemaPropertyForDomainObjectProperty } from './defineSqlSchemaPropertyForDomainObjectProperty';
import { isNotADatabaseGeneratedProperty } from './isNotADatabaseGeneratedProperty';

export const defineSqlSchemaRelationshipForDomainObject = ({
  domainObject,
  allDomainObjects,
}: {
  domainObject: DomainObjectMetadata;
  allDomainObjects: DomainObjectMetadata[];
}): SqlSchemaToDomainObjectRelationship => {
  // figure out some relevant info
  const sqlSchemaName = snakeCase(domainObject.name); // names are in snake case

  // define the properties
  const databaseGeneratedSqlSchemaProperties =
    defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject({
      domainObject,
    });
  const propertiesRelationship = [
    /// autogenerated values
    ...databaseGeneratedSqlSchemaProperties.map((sqlSchemaProperty) => {
      // see if this property was defined on the domain object (i.e., they want the value for it)
      const domainObjectProperty =
        Object.values(domainObject.properties).find(
          (property) => snakeCase(property.name) === sqlSchemaProperty.name,
        ) ?? null;

      // if so, validate that it has the correct type
      if (domainObjectProperty) {
        const expectedType = (() => {
          if (sqlSchemaProperty.name === 'id')
            return DomainObjectPropertyType.NUMBER;
          if (sqlSchemaProperty.name === 'uuid')
            return DomainObjectPropertyType.STRING;
          if (sqlSchemaProperty.name === 'created_at')
            return DomainObjectPropertyType.DATE;
          if (sqlSchemaProperty.name === 'effective_at')
            return DomainObjectPropertyType.DATE;
          if (sqlSchemaProperty.name === 'updated_at')
            return DomainObjectPropertyType.DATE;
          throw new UnexpectedCodePathDetectedError({
            reason:
              'unexpected database generated property name. this is a bug within sql-schema-generator.', // fail fast if expectations not met
            domainObjectName: domainObject.name,
            domainObjectPropertyName: sqlSchemaProperty.name,
          });
        })();
        if (domainObjectProperty.type !== expectedType)
          throw new UserInputError({
            reason: `reserved database-generated property does not match expected type. expected ${expectedType} but received ${domainObjectProperty.type}`,
            domainObjectName: domainObject.name,
            domainObjectPropertyName: domainObjectProperty.name,
          });
      }

      // return the relationship
      return {
        sqlSchema: sqlSchemaProperty,
        domainObject: domainObjectProperty,
      };
    }),

    // non autogenerated values
    ...Object.values(domainObject.properties)
      .filter(
        (domainObjectProperty) =>
          !databaseGeneratedSqlSchemaProperties
            .map(({ name }) => name)
            .includes(snakeCase(domainObjectProperty.name)), // filter out the db generated values, because we already got them above
      )
      .map((domainObjectProperty) => {
        return {
          sqlSchema: defineSqlSchemaPropertyForDomainObjectProperty({
            property: domainObjectProperty,
            domainObject,
            allDomainObjects,
          }),
          domainObject: domainObjectProperty,
        };
      }),
  ];

  // define the unique keys
  const sqlSchemaUniqueKeys = (() => {
    // define what the unique prop names are, based on what kind of object it is
    const domainObjectPropertyNamesInUniqueKey =
      domainObject.extends === DomainObjectVariant.DOMAIN_LITERAL
        ? propertiesRelationship // if its a literal, the schema is unique on all properties of the object (excluding autogenerated ones like id and uuid)
            .filter(isNotADatabaseGeneratedProperty)
            .map(
              ({ domainObject: domainObjectProperty }) =>
                domainObjectProperty.name,
            )
        : domainObject.decorations.unique; // otherwise, its unique on what was specified

    // convert the domain object prop names into sql schema prop names
    return (
      domainObjectPropertyNamesInUniqueKey?.map((domainObjectPropertyName) => {
        const sqlSchemaPropertyName = propertiesRelationship.find(
          (propertyRelationship) =>
            propertyRelationship.domainObject?.name ===
            domainObjectPropertyName,
        )?.sqlSchema.name;
        if (!sqlSchemaPropertyName)
          throw new UserInputError({
            reason: 'Unique keys must be properties of the domain object.',
            domainObjectName: domainObject.name,
            potentialSolution: ` Was there a typo defining '${
              domainObject.name
            }.unique' with key name '${domainObjectPropertyName}'? (Is it one of ${JSON.stringify(
              Object.keys(domainObject.properties),
            )}?)`,
          });
        return sqlSchemaPropertyName;
      }) ?? null
    );
  })();

  // make sure that this domain object is safe to have a schema defined for it
  assertDomainObjectIsSafeToHaveSqlSchema({ domainObject });

  // return the relationship
  return new SqlSchemaToDomainObjectRelationship({
    name: {
      sqlSchema: sqlSchemaName,
      domainObject: domainObject.name,
    },
    properties: propertiesRelationship,
    decorations: {
      alias: {
        domainObject: domainObject.decorations.alias,
      },
      unique: {
        sqlSchema: sqlSchemaUniqueKeys,
        domainObject: domainObject.decorations.unique,
      },
    },
  });
};
