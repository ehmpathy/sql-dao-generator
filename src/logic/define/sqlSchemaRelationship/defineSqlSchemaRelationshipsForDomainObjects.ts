import { DomainObjectMetadata } from 'domain-objects-metadata';

import { defineSqlSchemaRelationshipForDomainObject } from './defineSqlSchemaRelationshipForDomainObject';

/**
 * uses the full list of domain objects to define a relationship between its DomainObject representation and its SqlSchema representation
 */
export const defineSqlSchemaRelationshipsForDomainObjects = ({
  domainObjects,
}: {
  domainObjects: DomainObjectMetadata[];
}) =>
  domainObjects.map((domainObject) =>
    defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: domainObjects,
    }),
  );
