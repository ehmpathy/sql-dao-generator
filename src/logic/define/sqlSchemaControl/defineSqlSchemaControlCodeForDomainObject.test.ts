import {
  DomainObjectMetadata,
  DomainObjectPropertyType,
  DomainObjectVariant,
} from 'domain-objects-metadata';

import { defineSqlSchemaRelationshipForDomainObject } from '../sqlSchemaRelationship/defineSqlSchemaRelationshipForDomainObject';
import { defineSqlSchemaControlCodeForDomainObject } from './defineSqlSchemaControlCodeForDomainObject';

describe('defineSqlSchemaControlCodeForDomainObject', () => {
  it('should look right for a simple domain literal', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'Geocode',
      extends: DomainObjectVariant.DOMAIN_LITERAL,
      properties: {
        latitude: {
          name: 'latitude',
          type: DomainObjectPropertyType.NUMBER,
        },
        longitude: {
          name: 'longitude',
          type: DomainObjectPropertyType.NUMBER,
        },
      },
      decorations: {
        alias: null,
        primary: null,
        unique: null,
        updatable: null,
      },
    });
    const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: [domainObject],
    });

    // run it
    const code = defineSqlSchemaControlCodeForDomainObject({
      sqlSchemaRelationship,
    });

    // check that it looks right
    expect(code).toContain('path: ./tables/geocode.sql');
    expect(code).toContain('path: ./functions/upsert_geocode.sql');
    expect(code.split('\n').length).toEqual(7); // comment (1), static (2), upsert (2)
    expect(code).toMatchSnapshot();
  });
  it('should look right for a domain object with updatable properties', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'Carriage',
      extends: DomainObjectVariant.DOMAIN_ENTITY,
      properties: {
        id: {
          name: 'id',
          type: DomainObjectPropertyType.NUMBER,
          required: false,
        },
        uuid: {
          name: 'uuid',
          type: DomainObjectPropertyType.STRING,
          required: false,
        },
        cin: {
          name: 'cin',
          type: DomainObjectPropertyType.STRING,
          required: true,
        },
        carries: {
          name: 'carries',
          type: DomainObjectPropertyType.ENUM,
          of: ['PASSENGER', 'FREIGHT'],
          required: true,
        },
        capacity: {
          name: 'capacity',
          type: DomainObjectPropertyType.NUMBER,
          nullable: true,
        },
      },
      decorations: {
        alias: null,
        primary: null,
        unique: ['cin'],
        updatable: ['capacity'],
      },
    });
    const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: [domainObject],
    });

    // run it
    const code = defineSqlSchemaControlCodeForDomainObject({
      sqlSchemaRelationship,
    });

    // check that it looks right
    expect(code).toContain('path: ./tables/carriage.sql');
    expect(code).toContain('path: ./tables/carriage_version.sql');
    expect(code).toContain('path: ./tables/carriage_cvp.sql');
    expect(code).toContain('path: ./views/view_carriage_current.sql');
    expect(code).toContain('path: ./functions/upsert_carriage.sql');
    expect(code.split('\n').length).toEqual(13); // comment (1), resources (5x2)
    expect(code).toMatchSnapshot();
  });
  it('should look right for a domain object with static array properties', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'TrainLocatedEvent',
      extends: DomainObjectVariant.DOMAIN_EVENT,
      properties: {
        id: {
          name: 'id',
          type: DomainObjectPropertyType.NUMBER,
          required: false,
        },
        trainUuid: {
          name: 'trainUuid',
          type: DomainObjectPropertyType.STRING,
          required: true,
        },
        occurredAt: {
          name: 'occurredAt',
          type: DomainObjectPropertyType.DATE,
          required: true,
        },
        geocodes: {
          name: 'geocodes',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.REFERENCE,
            of: {
              name: 'Geocode',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
          required: true,
        },
        sensorUuids: {
          name: 'sensorUuids',
          type: DomainObjectPropertyType.ARRAY,
          of: { type: DomainObjectPropertyType.STRING },
        },
      },
      decorations: {
        alias: null,
        primary: null,
        unique: ['trainUuid', 'occurredAt'],
        updatable: [],
      },
    });
    const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: [
        domainObject,
        { name: 'Train', extends: DomainObjectVariant.DOMAIN_ENTITY },
      ] as DomainObjectMetadata[],
    });

    // run it
    const code = defineSqlSchemaControlCodeForDomainObject({
      sqlSchemaRelationship,
    });

    // check that it looks right
    expect(code).toContain('path: ./tables/train_located_event.sql');
    expect(code).toContain('path: ./tables/train_located_event_to_geocode.sql');
    expect(code).toContain(
      'path: ./tables/train_located_event_to_sensor_uuid.sql',
    );
    expect(code).toContain(
      'path: ./views/view_train_located_event_current.sql',
    );
    expect(code).toContain('path: ./functions/upsert_train_located_event.sql');
    expect(code.split('\n').length).toEqual(13); // comment (1), resources (5x2)
    expect(code).toMatchSnapshot();
  });
  it('should look right for a domain object with updatable array properties', () => {
    // define what we're testing on
    const domainObject = new DomainObjectMetadata({
      name: 'Train',
      extends: DomainObjectVariant.DOMAIN_ENTITY,
      properties: {
        id: {
          name: 'id',
          type: DomainObjectPropertyType.NUMBER,
          required: false,
        },
        uuid: {
          name: 'uuid',
          type: DomainObjectPropertyType.STRING,
          required: false,
        },
        badges: {
          name: 'badges',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.REFERENCE,
            of: {
              name: 'Badge',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
        },
        locomotiveUuids: {
          name: 'locomotiveUuids',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.STRING,
          },
        },
        leadEngineerUuid: {
          name: 'leadEngineerUuid',
          type: DomainObjectPropertyType.STRING,
        },
        sensorUuids: {
          name: 'sensorUuids',
          type: DomainObjectPropertyType.ARRAY,
          of: { type: DomainObjectPropertyType.STRING },
        },
      },
      decorations: {
        alias: null,
        primary: null,
        unique: ['uuid'],
        updatable: ['locomotiveUuids', 'leadEngineerUuid', 'sensorUuids'],
      },
    });
    const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: [
        domainObject,
        { name: 'Locomotive', extends: DomainObjectVariant.DOMAIN_ENTITY },
        { name: 'Engineer', extends: DomainObjectVariant.DOMAIN_ENTITY },
      ] as DomainObjectMetadata[],
    });

    // run it
    const code = defineSqlSchemaControlCodeForDomainObject({
      sqlSchemaRelationship,
    });

    // check that it looks right
    expect(code).toContain('path: ./tables/train.sql');
    expect(code).toContain('path: ./tables/train_to_badge.sql');
    expect(code).toContain('path: ./tables/train.sql');
    expect(code).toContain('path: ./tables/train_version_to_locomotive.sql');
    expect(code).toContain('path: ./tables/train_version_to_sensor_uuid.sql');
    expect(code).toContain('path: ./tables/train_cvp.sql');
    expect(code).toContain('path: ./views/view_train_current.sql');
    expect(code).toContain('path: ./functions/upsert_train.sql');
    expect(code.split('\n').length).toEqual(19); // comment (1), resources (8x2)
    expect(code).toMatchSnapshot();
  });
});
