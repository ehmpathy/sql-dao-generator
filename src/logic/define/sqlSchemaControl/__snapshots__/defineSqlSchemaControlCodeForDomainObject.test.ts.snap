// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineSqlSchemaControlCodeForDomainObject should look right for a domain object with static array properties 1`] = `
"# train_located_event
- type: resource
  path: ./tables/train_located_event.sql
- type: resource
  path: ./tables/train_located_event_to_geocode.sql
- type: resource
  path: ./tables/train_located_event_to_sensor_uuid.sql
- type: resource
  path: ./views/view_train_located_event_current.sql
- type: resource
  path: ./functions/upsert_train_located_event.sql"
`;

exports[`defineSqlSchemaControlCodeForDomainObject should look right for a domain object with updatable array properties 1`] = `
"# train
- type: resource
  path: ./tables/train.sql
- type: resource
  path: ./tables/train_to_badge.sql
- type: resource
  path: ./tables/train_version.sql
- type: resource
  path: ./tables/train_version_to_locomotive.sql
- type: resource
  path: ./tables/train_version_to_sensor_uuid.sql
- type: resource
  path: ./tables/train_cvp.sql
- type: resource
  path: ./views/view_train_current.sql
- type: resource
  path: ./functions/upsert_train.sql"
`;

exports[`defineSqlSchemaControlCodeForDomainObject should look right for a domain object with updatable properties 1`] = `
"# carriage
- type: resource
  path: ./tables/carriage.sql
- type: resource
  path: ./tables/carriage_version.sql
- type: resource
  path: ./tables/carriage_cvp.sql
- type: resource
  path: ./views/view_carriage_current.sql
- type: resource
  path: ./functions/upsert_carriage.sql"
`;

exports[`defineSqlSchemaControlCodeForDomainObject should look right for a simple domain literal 1`] = `
"# geocode
- type: resource
  path: ./tables/geocode.sql
- type: resource
  path: ./functions/upsert_geocode.sql"
`;
